design/KR/kenward.Rmd

```diff
@@ -0,0 +1,402 @@
+* `pbkrtest` implements the original Kenward-Roger degree of freedom instead of the improved one(recommended)
+* SAS software provides both Kenward-Roger degree of freedom
+* `lmerTestR` suggests that "The culprit is the calculation of the scaling of the $F$-value for which **pbkrtest** does not appear to export a low-level (direct) method."
+which leads to slow speed which we can improve
+
+# Prototypes
+
+## mathmatical details for KR1
+
+### adjusted variance-covariance matrix
+
+Following the vignettes' notations, we have
+\[
+  Y \sim N(X\beta, \Sigma)
+\]
+where $\Sigma$ is a block diagonal matrix.
+Let
+\[
+  \Phi(\sigma) = {\left\{X^\top \Sigma(\sigma)^\top X\right\}}^{-1}
+\]
+\[
+  \hat\beta = \Phi(\hat\sigma) X^\top \Sigma(\hat\sigma)^{-1} Y
+\]
+
+$\Phi$ is the variance-covariance matrix for $\hat\beta$, and we usually use
+$\hat\Phi = \Phi(\hat\sigma)$ to measure the precision of $\hat\beta$.
+
+However, Kackar and Harville shows that the variability can be further partitioned into
+\[
+  V(\hat\beta) = \Phi + \Lambda
+\]
+where $\Lambda$ represents the asymptotic variance-covariance matrix underestimates $V(\hat\beta)$,
+and $\Lambda$ can be approximated by
+\[
+  \Lambda \simeq \Phi \left\{\sum_{i=1}^r{\sum_{j=1}^r{W_{ij}(Q_{ij} - P_i \Phi P_j)} }\right\} \Phi
+\]
+where
+\[
+  P_i = X^\top \frac{\partial{\Sigma^{-1}}}{\partial \sigma_i} X
+\]
+\[
+  Q_{ij} = X^\top \frac{\partial{\Sigma^{-1}}}{\partial \sigma_i} \Sigma \frac{\partial{\Sigma^{-1}}}{\partial \sigma_j} X
+\]
+and $W_{ij}$ is the $(i, j)$ element of $V(\hat\sigma)$.
+
+Further consider the bias of $\hat\Phi$ as an estimator of $\Phi$ using Taylor expansion
+
+\[
+  \hat\Phi \simeq \Phi + \sum_{i=1}^r{(\hat\sigma_i - \sigma_i)\frac{\partial{\Phi}}{\partial{\sigma_i}}} + \frac{1}{2} \sum_{i=1}^r{\sum_{j=1}^r{(\hat\sigma_i - \sigma_i)(\hat\sigma_j - \sigma_j)\frac{\partial^2{\Phi}}{\partial{\sigma_i}\partial{\sigma_j}}}}
+\]
+
+Ignoring the bias and the expectation is
+\[
+  E(\hat\Phi) = \Phi + \frac{1}{2} \sum_{i=1}^r{\sum_{j=1}^r{W_{ij}\frac{\partial^2{\Phi}}{\partial{\sigma_i}\partial{\sigma_j}}}}
+\]
+
+Using previous notations,
+
+\[
+  \frac{\partial^2{\Phi}}{\partial{\sigma_i}\partial{\sigma_j}} = \Phi (P_i \Phi P_j + P_j \Phi P_i - Q_{ij} - Q_{ji} + R_{ij}) \Phi
+\]
+where
+\[
+  R_{ij} = X^\top\Sigma^{-1}\frac{\partial^2\Sigma}{\partial{\sigma_i}\partial{\sigma_j}} \Sigma^{-1} X
+\]
+
+(Note: in some implementations of KR, $R_{ij}$ is ignored and leads to a variant.)
+
+So the adjusted estimator is
+
+\[
+  \hat\Phi_A = \hat\Phi + 2\hat\Phi \left\{\sum_{i=1}^r{\sum_{j=1}^r{W_{ij}(Q_{ij} - P_i \hat\Phi P_j - \frac{1}{4}R_{ij})} }\right\} \hat\Phi
+\]
+
+$W$ can be estimated using the inverse of information matrix by substituting the observed values
+
+\[
+  2W_{ij} = tr(\frac{\partial{\Sigma^{-1}}}{\partial \sigma_i} \Sigma \frac{\partial{\Sigma^{-1}}}{\partial \sigma_j} \Sigma) - tr(2\Phi Q_{ij} - \Phi P_i\Phi P_j)
+\]
+
```

Comment:
So before jumping into coding I would like to see here more simplifications and explanations / thoughts on the adjusted cov estimator, e.g.
- explain how Sigma is block diagonal into subjects
- explain that we have the derivatives of Sigma blocks exactly based on the chosen covariance structure
- think about how we will get these exact derivatives, I think we need to do that in C++, question is whether we can use the automatic differentiation engine within TMB to avoid manual coding of this?
- explain we have W
- break down the P, Q, R calculations into sums of subject specific terms where possible (hopefully possible for all)

Reply:
@clarkliming I would start with this steps here